openssl-sys/build.rs

```diff
@@ -79,6 +74,137 @@ fn main() {
     }
 }
 
+fn build_openssl(target: &str, tarball_path: &Path) -> (PathBuf, PathBuf) {
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let build_dir = out_dir.join("build");
+    let install_dir = out_dir.join("install");
+    let lib_dir = install_dir.join("lib");
+    let include_dir = install_dir.join("include");
+
+    let stamp = build_dir.join("stamp");
+    let mut contents = String::new();
+    let _ = File::open(&stamp).and_then(|mut f| f.read_to_string(&mut contents));
+    if tarball_path == Path::new(&contents) {
+        return (lib_dir, include_dir);
+    }
+
+    if stamp.exists() {
+        fs::remove_file(&stamp).unwrap();
+    }
+    if build_dir.exists() {
+        fs::remove_dir_all(&build_dir).unwrap();
+    }
+    if install_dir.exists() {
+        fs::remove_dir_all(&install_dir).unwrap();
+    }
+
+    let tarball = File::open(tarball_path).unwrap().gz_decode().unwrap();
+    let mut tarball = Archive::new(tarball);
+    tarball.unpack(&build_dir).unwrap();
+
+    let inner_dir = fs::read_dir(&build_dir).unwrap().next().unwrap().unwrap().path();
+
+    let mut configure = if target == "i686-unknown-linux-gnu" {
+        let mut cmd = Command::new("setarch");
+        cmd.arg("i386").arg("./Configure");
+        cmd
+    } else {
+        Command::new("./Configure")
+    };
+
+    let os = match target {
+        "aarch64-unknown-linux-gnu" => "linux-aarch64",
+        "arm-unknown-linux-gnueabi" => "linux-armv4",
+        "arm-unknown-linux-gnueabihf" => "linux-armv4",
+        "armv7-unknown-linux-gnueabihf" => "linux-armv4",
+        "i686-apple-darwin" => "darwin-i386-cc",
+        "i686-unknown-freebsd" => "BSD-x86-elf",
+        "i686-unknown-linux-gnu" => "linux-elf",
+        "i686-unknown-linux-musl" => "linux-elf",
+        "mips-unknown-linux-gnu" => "linux-mips32",
+        "mips64-unknown-linux-gnuabi64" => "linux64-mips64",
+        "mips64el-unknown-linux-gnuabi64" => "linux64-mips64",
+        "mipsel-unknown-linux-gnu" => "linux-mips32",
+        "powerpc-unknown-linux-gnu" => "linux-ppc",
+        "powerpc64-unknown-linux-gnu" => "linux-ppc64",
+        "powerpc64le-unknown-linux-gnu" => "linux-ppc64le",
+        "s390x-unknown-linux-gnu" => "linux64-s390x",
+        "x86_64-apple-darwin" => "darwin64-x86_64-cc",
+        "x86_64-unknown-freebsd" => "BSD-x86_64",
+        "x86_64-unknown-linux-gnu" => "linux-x86_64",
+        "x86_64-unknown-linux-musl" => "linux-x86_64",
+        "x86_64-unknown-netbsd" => "BSD-x86_64",
+        _ => panic!("don't know how to configure OpenSSL for {}", target),
+    };
+
+    configure.arg(format!("--prefix={}", install_dir.display()))
+        .arg("no-dso")
+        .arg("no-ssl2")
+        .arg("no-ssl3")
+        .arg("no-comp")
+        .arg(os)
+        .arg("-fPIC");
+    if target.contains("i686") {
+        configure.arg("-m32");
```

Comment:
You may want to read the cflags from gcc-rs to pass here (instead of manually passing -fPIC/-m32)